### PR TITLE
Bug fixes, better error messages, and tests for Tridiag, Bidiag, and BLAS

### PR DESCRIPTION
**Pull request by kshyatt**
Some of the code in `tridiag.jl` was obviated by `special.jl`. I checked that it was not called and removed it. I added a missing convert method to `special.jl`. Hopefully this gets `tridiag.jl` to 100% coverage!

BLAS gets a bunch more tests and more verbose error messages. I split some of them up for coverage purposes.

The `convert` method from `Bidiagonal` to `Tridiagonal` was broken. I fixed it and added tests for `promote_rule` and now `Bidiag` has all methods tested too!

### CODE REVIEW

**Comment by kshyatt** [comment]
One of the eigenvectors has flipped sign on `i686`. @andreasnoack, any ideas what's up?

**Comment by kshyatt** [comment]
I also added tests for `gbmv`, `sbmv`, and implemented and tested `hbmv`. `sbmv` DOES NOT work on complex numbers!

**Comment by andreasnoack** [comment]
I'm sure why the vectors match exactly on 64 bit, but have a sign flipped on 32 bit. The elementary reflectors applied to the `full` version of the matrix flip the sign of the off diagonal elements which could possible affect the calculations, but it doesn't seem to be the case on 64 bit.

File: base/linalg/blas.jl [hunk 1]
```diff
@@ -848,7 +895,9 @@ for (mmname, smname, elty) in
                        A::StridedMatrix{$elty}, B::StridedMatrix{$elty})
             m, n = size(B)
             nA = chksquare(A)
-            if nA != (side == 'L' ? m : n) throw(DimensionMismatch("trmm!")) end
+            if nA != (side == 'L' ? m : n)
+                throw(DimensionMismatch("Size of A, $(size(A)), doesn't match $side size of B with dims, $(size(B))"))
```
Comment (KristofferC) [inline base/linalg/blas.jl]: I think no capitalization here.

Comment (kshyatt) [inline base/linalg/blas.jl]: Every time, I make this mistake. Thanks :)

**Comment by andreasnoack** [comment]
Did you change anything in the test of the eigenvectors?

**Comment by kshyatt** [comment]
I switched to `test_approx_eq_modphase`

File: deps/libmojibake [hunk 2]
```diff
@@ -0,0 +1 @@
+Subproject commit 86447ad060d6f4edf01f2a64b9598dfeeb6e6f7d
```
Comment (andreasnoack) [inline deps/libmojibake]: I don't think this change belongs here

Comment (kshyatt) [inline deps/libmojibake]: Yeah sorry git was being dumb, will remove.

File: base/linalg/special.jl [hunk 3]
```diff
@@ -70,6 +70,10 @@ function convert(::Type{SymTridiagonal}, A::Tridiagonal)
     SymTridiagonal(A.d, A.dl)
 end
 
+function convert(::Type{Tridiagonal}, A::SymTridiagonal)
+    Tridiagonal(A.ev, A.dv, A.ev)
```
Comment (andreasnoack) [inline base/linalg/special.jl]: I think we might want to avoid the aliasing between the two off-diagonals here.

**Comment by StefanKarpinski** [comment]
> I switched to `test_approx_eq_modphase`

I find the explosion of weird approximate equality test helpers kind of unsettling. Would it make sense to have specialized kinds of approximate equality for different kinds of arguments? We can already write `@test A ≈ B` instead of `@test_approx_eq A B`; could we also write that same thing and have it mean whatever's appropriate for testing approximate equality in these circumstances?

**Comment by andreasnoack** [comment]
This is a great cleanup (and test improvement, of course).

I don't know which solution is the best for test macros. It is convenient with respect to error messages that the macro takes all the components of the test as arguments, but the cost seems to be that we need more test macros. However, there might be good solutions I haven't thought of. "Overloading" ≈ is difficult in the case of eigenvectors because only we know that they are only determined up to phase.

**Comment by jiahao** [comment]
`test_approx_eq_modphase` (ref #10651) tests a very specific issue, that eigenvectors and singular vectors are only defined up to an overall sign (or phase in the complex case). In exact arithmetic one can formally write the test as "a and b are in the same equivalence class modulo U(1)" (U(1) being the group of complex phases). The problem is that there isn't really standard notation that is easily parsable that means "approximately equal modulo U(1)". ≅?

**Comment by andreasnoack** [comment]
I tried to find ≈ with a circle on top, but it doesn't appear to exist.

**Comment by carnaval** [comment]
You could express it as a quotient `MulMod` type such that
```julia
==(x::MulMod{T},y::MulMod{T}) = x/y in T
```
and define a type that represents the complex unit circle.

But that may be slightly overkill :p

**Comment by kshyatt** [comment]
Unless there are other needed changes, I'm going to merge.

**Pull request merged**